File: clojure/cljsearch/test/cljsearch/searchresult_test.clj
(ns cljsearch.searchresult-test
  (:require [clojure.test :refer :all])
  (:use [clojure.string :as str :only (trim)]
        [clojure.java.io :only (file)]
        [cljsearch.config :only (XSEARCHPATH)]
        [cljfind.consolecolor]
        [cljfind.fileresult :only (new-file-result file-result-path)]
        [cljfind.fileutil :only (get-path-name to-path)]
        [cljsearch.searchresult :only (->SearchResult get-search-result-formatter)]
        [cljsearch.searchsettings :only (DEFAULT-SEARCH-SETTINGS)]))

(def CSSEARCHPATH
  (str XSEARCHPATH "/csharp/CsSearch/CsSearch"))

(deftest test-singleline-search-result
  (testing "test-singleline-search-result"
    (let [settings (assoc DEFAULT-SEARCH-SETTINGS :colorize false)
          file (to-path (str CSSEARCHPATH "/Searcher.cs"))
          file-result (new-file-result file :code)
          line-num 10
          match-start-index 15
          match-end-index 23
          line "\tpublic class Searcher\n"
          lines-before []
          lines-after []
          r (->SearchResult
              (re-pattern "Search")
              file-result
              line-num
              match-start-index
              match-end-index
              line
              lines-before
              lines-after)
          expected (format "%s: %d: [%d:%d]: %s" (file-result-path file-result) line-num
            match-start-index match-end-index (str/trim line))
          format-search-result (get-search-result-formatter settings)]
      (is (= (format-search-result r) expected)))))

(deftest test-singleline-longer-than-maxlength-search-result
  (testing "test-singleline-longer-than-maxlength-search-result"
    (let [settings (assoc DEFAULT-SEARCH-SETTINGS :colorize false :max-line-length 100)
          file (to-path "./maxlen.txt")
          file-result (new-file-result file :text)
          line-num 1
          match-start-index 53
          match-end-index 59
          line "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
          lines-before []
          lines-after []
          r (->SearchResult
              (re-pattern "maxlen")
              file-result
              line-num
              match-start-index
              match-end-index
              line
              lines-before
              lines-after)
          expected-line "...89012345678901234567890123456789012345678901maxlen89012345678901234567890123456789012345678901..."
          expected (format "%s: %d: [%d:%d]: %s" (file-result-path file-result) line-num
            match-start-index match-end-index expected-line)
          format-search-result (get-search-result-formatter settings)]
      (is (= (format-search-result r) expected)))))

(deftest test-singleline-colorize-search-result
  (testing "test-singleline-longer-than-maxlength-search-result"
    (let [settings (assoc DEFAULT-SEARCH-SETTINGS :colorize true :max-line-length 100)
          file (to-path "./maxlen.txt")
          file-result (new-file-result file :text)
          line-num 1
          match-start-index 53
          match-end-index 59
          line "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
          lines-before []
          lines-after []
          r (->SearchResult
              (re-pattern "maxlen")
              file-result
              line-num
              match-start-index
              match-end-index
              line
              lines-before
              lines-after)
          expected-line (str "...89012345678901234567890123456789012345678901" CONSOLE_COLOR_GREEN "maxlen" CONSOLE_COLOR_RESET "89012345678901234567890123456789012345678901...")
          expected (format "%s: %d: [%d:%d]: %s" (file-result-path file-result) line-num
            match-start-index match-end-index expected-line)
          format-search-result (get-search-result-formatter settings)]
      (is (= (format-search-result r) expected)))))

(deftest test-binaryfile-search-result
  (testing "test-binaryfile-search-result"
    (let [settings DEFAULT-SEARCH-SETTINGS
          file (to-path (str CSSEARCHPATH "/Searcher.exe"))
          file-result (new-file-result file :binary)
          line-num 0
          match-start-index 0
          match-end-index 0
          line nil
          lines-before []
          lines-after []
          r (->SearchResult
              (re-pattern "Search")
              file-result
              line-num
              match-start-index
              match-end-index
              line
              lines-before
              lines-after)
          expected (format "%s matches at [%d:%d]" (file-result-path file-result) match-start-index
            match-end-index)
          format-search-result (get-search-result-formatter settings)]
      (is (= (format-search-result r) expected)))))

(deftest test-multiline-search-result
  (testing "test-multiline-search-result"
    (let [settings DEFAULT-SEARCH-SETTINGS
          file (to-path (str CSSEARCHPATH "/Searcher.cs"))
          file-result (new-file-result file :code)
          line-num 10
          match-start-index 15
          match-end-index 23
          line "\tpublic class Searcher\n"
          lines-before ["namespace CsSearch\n" "{\n"]
          lines-after ["\t{\n" "\t\tprivate readonly FileTypes _fileTypes;\n"]
          r (->SearchResult
              (re-pattern "Search")
              file-result
              line-num
              match-start-index
              match-end-index
              line
              lines-before
              lines-after)
          outputformat
"================================================================================
%s: %d: [%d:%d]
--------------------------------------------------------------------------------
   8 | namespace CsSearch
   9 | {
> 10 | \tpublic class Searcher
  11 | \t{
  12 | \t\tprivate readonly FileTypes _fileTypes;
"
          expected (format outputformat (file-result-path file-result) line-num
            match-start-index match-end-index)
          format-search-result (get-search-result-formatter settings)]
      (is (= (format-search-result r) expected)))))

File: cpp/cppsearch/tests/SearchResultTest.cpp
#include <catch2/catch_all.hpp>
#include "SearchFileResult.h"
#include "SearchResultFormatter.h"
#include "SearchSettings.h"
#include "cppfind.h"

TEST_CASE("Verify single-line result equals expected", "[SearchResult]") {
    auto settings = cppsearch::SearchSettings();
    settings.colorize(false);
    auto formatter = cppsearch::SearchResultFormatter(settings);
    auto pattern = cppfind::RegexPattern("Search");

    const std::filesystem::path file_path{"~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"};
    auto file_type = cppfind::FileType::CODE;
    uint64_t file_size = 1000;
    long mod_time = 1000;
    auto file_result = cppfind::FileResult(file_path, file_type, file_size, mod_time);

    unsigned long line_num = 10;
    unsigned long match_start_idx = 15;
    unsigned long match_end_idx = 23;
    std::string line = "\tpublic class Searcher\n";
    auto text_result = cppsearch::SearchTextResult(pattern, line_num, match_start_idx, match_end_idx, line);
    auto result = cppsearch::SearchFileResult(file_result, std::move(text_result));
    std::string expected_path = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs";
    std::string expected_line = "public class Searcher";
    std::string expected_output = expected_path + ": " + std::to_string(line_num) + ": ["
            + std::to_string(match_start_idx) + ":" + std::to_string(match_end_idx) + "]: " + expected_line;
    std::string output = formatter.format(result);

    REQUIRE(output == expected_output);
}

TEST_CASE("Verify single-line-longer-than-max_line_length result equals expected", "[SearchResult]") {
    auto settings = cppsearch::SearchSettings();
    settings.colorize(false);
    settings.max_line_length(100);
    auto formatter = cppsearch::SearchResultFormatter(settings);
    auto pattern = cppfind::RegexPattern("maxlen");

    std::filesystem::path file_path{"./maxlen.txt"};
    auto file_type = cppfind::FileType::TEXT;
    uint64_t file_size = 0;
    long mod_time = 0;
    auto file_result = cppfind::FileResult(std::move(file_path), file_type, file_size, mod_time);

    unsigned long line_num = 1;
    unsigned long match_start_idx = 53;
    unsigned long match_end_idx = 59;
    std::string line = "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789";
    auto text_result = cppsearch::SearchTextResult(pattern, line_num, match_start_idx, match_end_idx, line);
    auto result = cppsearch::SearchFileResult(file_result, std::move(text_result));
    std::string expected_path = "./maxlen.txt";
    std::string expected_line = "...89012345678901234567890123456789012345678901maxlen89012345678901234567890123456789012345678901...";
    std::string expected_output = expected_path + ": " + std::to_string(line_num) + ": ["
            + std::to_string(match_start_idx) + ":" + std::to_string(match_end_idx) + "]: " + expected_line;
    std::string output = formatter.format(result);

    REQUIRE(output == expected_output);
}

TEST_CASE("Verify single-line-longer-and-colorize result equals expected", "[SearchResult]") {
    auto settings = cppsearch::SearchSettings();
    settings.colorize(true);
    settings.max_line_length(100);
    auto formatter = cppsearch::SearchResultFormatter(settings);
    auto pattern = cppfind::RegexPattern("maxlen");

    std::filesystem::path file_path{"./maxlen.txt"};
    auto file_type = cppfind::FileType::TEXT;
    uint64_t file_size = 0;
    long mod_time = 0;
    auto file_result = cppfind::FileResult(std::move(file_path), file_type, file_size, mod_time);

    unsigned long line_num = 1;
    unsigned long match_start_idx = 53;
    unsigned long match_end_idx = 59;
    std::string line = "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789";
    auto text_result = cppsearch::SearchTextResult(pattern, line_num, match_start_idx, match_end_idx, line);
    auto result = cppsearch::SearchFileResult(file_result, std::move(text_result));
    std::string expected_path = "./maxlen.txt";
    std::string expected_line = std::string("...89012345678901234567890123456789012345678901") +
            COLOR_GREEN +
            "maxlen" +
            COLOR_RESET +
            "89012345678901234567890123456789012345678901...";
    std::string expected_output = expected_path + ": " + std::to_string(line_num) + ": ["
            + std::to_string(match_start_idx) + ":" + std::to_string(match_end_idx) + "]: " + expected_line;
    std::string output = formatter.format(result);

    REQUIRE(output == expected_output);
}

TEST_CASE("Verify binary result equals expected", "[SearchResult]") {
    auto settings = cppsearch::SearchSettings();
    auto formatter = cppsearch::SearchResultFormatter(settings);
    auto pattern = cppfind::RegexPattern("Search");

    std::filesystem::path file_path{"~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.exe"};
    auto file_type = cppfind::FileType::BINARY;
    uint64_t file_size = 0;
    long mod_time = 0;
    auto file_result = cppfind::FileResult(std::move(file_path), file_type, file_size, mod_time);

    unsigned long line_num = 0;
    unsigned long match_start_idx = 0;
    unsigned long match_end_idx = 0;
    auto text_result = cppsearch::SearchTextResult(pattern, line_num, match_start_idx, match_end_idx, "");
    auto result = cppsearch::SearchFileResult(file_result, std::move(text_result));
    std::string expected_path = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.exe";
    std::string expected_output = expected_path + " matches at [0:0]";
    std::string output = formatter.format(result);

    REQUIRE(output == expected_output);
}

TEST_CASE("Verify multiline result equals expected", "[SearchResult]") {
    auto settings = cppsearch::SearchSettings();
    settings.colorize(false);
    settings.lines_before(2);
    settings.lines_after(2);
    auto formatter = cppsearch::SearchResultFormatter(settings);
    auto pattern = cppfind::RegexPattern("Search");

    std::filesystem::path file_path{"~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"};
    auto file_type = cppfind::FileType::CODE;
    uint64_t file_size = 0;
    long mod_time = 0;
    auto file_result = cppfind::FileResult(std::move(file_path), file_type, file_size, mod_time);
    unsigned long line_num = 10;
    unsigned long match_start_idx = 15;
    unsigned long match_end_idx = 23;
    std::string line = "\tpublic class Searcher";
    std::vector<std::string> lines_before = {"namespace CsSearch", "{"};
    std::vector<std::string> lines_after = {"\t{", "\t\tprivate readonly FileTypes _fileTypes;"};
    auto text_result = cppsearch::SearchTextResult(pattern, line_num, match_start_idx, match_end_idx, line,
        lines_before, lines_after);
    auto result = cppsearch::SearchFileResult(file_result, std::move(text_result));
    std::string expected_path = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs";
    std::string expected_output = std::string("================================================================================\n") +
            expected_path + ": " + std::to_string(line_num) + ": [" + std::to_string(match_start_idx) + ":" + std::to_string(match_end_idx) + "]\n" +
            "--------------------------------------------------------------------------------\n" +
            "   8 | namespace CsSearch\n" +
            "   9 | {\n" +
            "> 10 | \tpublic class Searcher\n" +
            "  11 | \t{\n" +
            "  12 | \t\tprivate readonly FileTypes _fileTypes;\n";
    std::string output = formatter.format(result);

    REQUIRE(output == expected_output);
}
